modules/internal/ArrayViewRankChange.chpl

```diff
@@ -22,9 +22,337 @@
 // This module provides an array domain map class that is used to
 // represent rank change slices of another array via a domain.
 //
+// Throughout this file, 'down*' refers to the higher-dimensional
+// array/domain/distribution that's been rank-changed in this view,
+// while 'up*' refers to the lower-dimensional, user-facing view.
+//
+// TODOs for this file:
+//
+// * the dist, dom, and arr classes store a certain amount of
+//   duplicated information -- should these be focused in the higher
+//   classes (dist/dom) and referenced by lower ones (dom/dist)?
+//
+// * I don't think I've done anything to privatize the dist or dom
+//   classes... should I?
+//
 module ArrayViewRankChange {
 
   //
+  // This class represents a distribution that knows how to create
+  // rank-change domains and arrays similar to the one that caused
+  // it to be created.  It stores a pointer down to the distribution
+  // that it creates lower-dimensional views of, and two tuples
+  // indicating which dimensions were collapsed by the rank-change
+  // and what the indices of those dimensions are.
+  //
+  class ArrayViewRankChangeDist: BaseDist {
+    const downdist;
+    const collapsedDim;
+    const idx;
+
+    proc dsiNewRectangularDom(param rank, type idxType, param stridable, inds) {
+      var newdom = new ArrayViewRankChangeDom(rank=rank,
+                                              idxType=idxType,
+                                              stridable=stridable,
+                                              collapsedDim=collapsedDim,
+                                              idx=idx,
+                                              dist=this);
+      newdom.dsiSetIndices(inds);
+      return newdom;
+    }
+
+    proc dsiClone() return new ArrayViewRankChangeDist(downdist=downdist,
+                                                       collapsedDim=collapsedDim,
+                                                       idx=idx);
+  }
+
+  //
+  // This class represents the domain of a rank-change slice of an
+  // array.  Like other domain class implementations, it supports the
+  // standard dsi interface.  Note that rank changes only make sense
+  // for rectangular domains so this is a subclass of
+  // BaseRectangularDom.
+  //
+ class ArrayViewRankChangeDom: BaseRectangularDom {
+    param rank;
+    type idxType;
+    param stridable;
+
+    // the lower-dimensional index set that we represent upwards
+    var updom: DefaultRectangularDom(rank, idxType, stridable);
+
+    // the collapsed dimensions and indices in those dimensions
```

Comment:
I had a general TODO to this effect at the top of the file but have added an additional one here.  The main reason I didn't do it is because, by the time I got everything working, leak-free, and performing well, I was too scared to change anything else for fear of breaking something -- most likely performance (this is a pretty nervous reason, though... keep in mind that this PR was finally "ready" at 11:40pm last night... :)  ).